include/revng/Model/Helpers.h

```diff
@@ -12,23 +12,59 @@
 namespace model {
 
 template<typename Type>
-concept EntityWithMetadata = requires(Type &&Value) {
+concept EntityWithCustomName = requires(Type &&Value) {
   { Value.CustomName() } -> convertible_to<model::Identifier &>;
+};
+
+template<typename Type>
+concept EntityWithOriginalName = requires(Type &&Value) {
   { Value.OriginalName() } -> convertible_to<std::string &>;
 };
 
-template<EntityWithMetadata LHS, EntityWithMetadata RHS>
+template<typename Type>
+concept EntityWithComment = requires(Type &&Value) {
+  { Value.Comment() } -> convertible_to<std::string &>;
+};
+
+template<typename Type>
+concept EntityWithReturnValueComment = requires(Type &&Value) {
+  { Value.ReturnValueComment() } -> convertible_to<std::string &>;
+};
+
+template<typename LHS, typename RHS>
 LHS &copyMetadata(LHS &To, const RHS &From) {
-  To.CustomName() = From.CustomName();
-  To.OriginalName() = From.OriginalName();
+  if constexpr (EntityWithCustomName<LHS> && EntityWithCustomName<RHS>)
+    To.CustomName() = From.CustomName();
+
+  if constexpr (EntityWithOriginalName<LHS> && EntityWithOriginalName<RHS>)
+    To.OriginalName() = From.OriginalName();
+
+  if constexpr (EntityWithComment<LHS> && EntityWithComment<RHS>)
+    To.Comment() = From.Comment();
+
+  if constexpr (EntityWithReturnValueComment<LHS>
+                && EntityWithReturnValueComment<RHS>) {
+    To.ReturnValueComment() = From.ReturnValueComment();
+  }
 
   return To;
 }
```

Comment:
1. Very enterprise.
2. `static_assert(EntityWithCustomName<LHS> == EntityWithCustomName<RHS>)` and friends.
3. I see why you're sad about `ReturnValueComment`. An alternative would be to have a trait `copyMetadata` that we can extend specifically for those with `ReturnValueComment`, but your solution works too.

Reply:
1. Glad you like it.

2. What do you mean? It's intentional that these two have different types. For example, this is used for `MyRFT.CustomName() = MyCFT.CustomName()` and so on you, you get the point. The idea is, as long as BOTH sides have the field in question, copy/move the value.

3. Yeah, `ReturnValueComment` sticks as a sore thumb. But I can't think of anything better. To the point where I even abandoned the effort of encoding comments for each register separately in RFT and went with the same approach.

Reply:
2. OK
3. >  I even abandoned the effort of encoding comments for each register separately in RFT and went with the same 

   What do you mean? How are we encoding comments of RFT args?

Reply:
3. Sorry, I mean for return values. Arguments work as you would expect.

Reply:
Uhm, what's difficult about RFT return values? Can't we add `Comment` to `TypedRegister` and then concat them upon emission similarly to what we decided to do for `StackArgumentsType`?

Reply:
Yeah, I abandoned that idea initially, but since we're now getting back to it, this might have some merit.

The main problem is how to encapsulate it into the type conversion. With stack it simple since we'll have a new argument for each field - so the comments are passed through in the same way as the rest of the metadata. In CFT it doesn't make sense to have more than a single comment for the return value. I guess we could make this "concatenation of multiple comments" thing generic enough to be used everywhere, but I have a lot of FUD about that.

Reply:
Concatenation is fine.

Just make sure you wrap each thing in a proper [PTML `data-model-edit-attribute` tag](https://pad.rev.ng/s/L3m9LemOh) so that users can edit comments directly from the UI.

Reply:
Oh, right, thank you for the remainder!

Reply:
> Concatenation is fine.

I have a feeling like we already had this conversation I'm **really** hesitant to use any non-reversible transformations when converting user input (which comments are). Doing something stupid like going `RFT->CFT->RFT->CFT->RFT` could get really messy if we have some non-trivial comment conversions in the mix, especially once we would have different versions of the same analysis in the wild, which introduces possibilities of some of these conversions done differently from the others.